sdk/src/test/java/com/qonversion/android/sdk/internal/user/controller/UserControllerTest.kt

```diff
@@ -305,4 +318,187 @@ internal class UserControllerTest {
             verify(exactly = 1) { mockLogger.error("Failed to update user cache", exception) }
         }
     }
+
+    @Nested
+    inner class OnAppForegroundTest {
+
+        @Test
+        fun `not the first foreground`() {
+            // given
+
+            // when
+            userController.onAppForeground(false)
+
+            // then
+            verify {
+                mockUserCacher wasNot called
+                mockUserService wasNot called
+            }
+        }
+
+        @Test
+        fun `actual cache exists`() = runTest {
+            // given
+            userController = createController(this)
+            every { mockUserCacher.getActual() } returns mockUser
+
+            // when
+            userController.onAppForeground(true)
+
+            // then
+            yield()
+            verify { mockUserCacher.getActual() }
+            verify(exactly = 0) {
+                mockUserCacher.store(any())
+                mockUserService wasNot called
+            }
+        }
+
+        @Test
+        fun `actual cache does not exist`() = runTest {
+            // given
+            userController = spyk(createController(this))
+            every { mockUserCacher.getActual() } returns null
+            every { mockUserDataStorage.requireUserId() } returns testUserId
+            coEvery { mockUserService.getUser(testUserId) } returns mockUser
+            coEvery { userController.handleNewUserInfo(mockUser) } just runs
+
+            // when
+            userController.onAppForeground(true)
+
+            // then
+            yield()
+            coVerifyOrder {
+                mockUserCacher.getActual()
+                mockUserDataStorage.requireUserId()
+                mockUserService.getUser(testUserId)
+                userController.handleNewUserInfo(mockUser)
+            }
+        }
+
+        @Test
+        fun `cacher throws exception`() = runTest {
+            // given
+            val exception = QonversionException(ErrorCode.BackendError)
+            userController = spyk(createController(this))
+            every { mockUserCacher.getActual() } throws exception
+            coEvery { userController.handleNewUserInfo(mockUser) } just runs
+
+            // when
+            assertDoesNotThrow {
+                userController.onAppForeground(true)
+            }
+
+            // then
+            yield()
+            verify {
+                mockUserCacher.getActual()
+                mockLogger.error("Requesting user on app first foreground failed", exception)
+            }
+            coVerify(exactly = 0) {
+                mockUserDataStorage wasNot called
+                mockUserService wasNot called
+                userController.handleNewUserInfo(mockUser)
+            }
+        }
+    }
```

Comment:
Let's also add three tests when: `userDataStorage.requireUserId()`, `userService.getUser(userId) ` and `handleNewUserInfo(newUser)` throw exceptions